### PR TITLE
fix combox and container lua widget behaviour

### PR DESCRIPTION
**Pull request by theres**
Generally speaking, comboboxes & containers now appears more like standard lua tables:
- access to numeric element out-of-bound return `nil`
- access to children of container by index is working (fix regression from 0edd2f429b905e86ce5dbc6cbd3db1beea5d4c1f)

After this changes:
- it is possible to get container (box, stack) element by index
- it is possible to use `for` with `pair` to list all attributes and children of containers & comboboxes
- it is possible to use `for` with `iparis` to list all children of containers & comboboxes

some examples of what this do: [tests.lua.txt](https://github.com/darktable-org/darktable/files/1066710/tests.lua.txt)

### CODE REVIEW

**Comment by LebedevRI** [comment]
Does this some issue listed in redmine?
CC @boucman

**Comment by theres** [comment]
I've not found any redmine issue for this. Shoud I create one?

File: doc/usermanual/lua/lua_api.xml [hunk 1]
```diff
@@ -7564,7 +7564,7 @@ This function is recursion-safe and can be used to dump _G if needed.</para>
 <para>The widgets contained by the box</para>
 <para>You can append widgets by adding them at the end of the list</para>
 <para>You can remove widgets by setting them to nil</para>
-
+<para>You can replace widgets at specific index by simply assignment</para>
```
Comment (houz) [inline doc/usermanual/lua/lua_api.xml]: I'd reword it like this: "You can replace widgets at a specific index by a simple assignment".

File: src/lua/widget/container.c [hunk 2]
```diff
@@ -154,16 +154,35 @@ static int container_numindex(lua_State*L)
       GtkWidget *searched_widget = g_list_nth_data(children,index);
       gtk_container_remove(GTK_CONTAINER(container->widget),searched_widget);
     } else {
-      luaL_error(L,"Incorrect index or value when setting the child of a container : you can only append, not change a child in the middle");
+      GList *current;
+      for(current = g_list_nth(children, index); current != NULL; current = current->next)
+      {
+        gtk_container_remove(GTK_CONTAINER(container->widget), current->data);
```
Comment (houz) [inline src/lua/widget/container.c]: `gtk_container_remove()` may destroy the widget, so call `g_object_ref()` on the widget first.

File: src/lua/widget/container.c [hunk 3]
```diff
@@ -154,16 +154,35 @@ static int container_numindex(lua_State*L)
...
+      }
+
+      lua_widget widget;
+      luaA_to(L, lua_widget, &widget, 3);
+      gtk_container_add(GTK_CONTAINER(container->widget), widget->widget);
+
+      for(current = g_list_nth(children, index + 1); current != NULL; current = current->next)
+      {
+        gtk_container_add(GTK_CONTAINER(container->widget), current->data);
```
Comment (houz) [inline src/lua/widget/container.c]: After adding the old widget back you have to `g_object_unref()` it again.

Comment (theres) [inline src/lua/widget/container.c]: done

File: src/lua/widget/container.c [hunk 4]
```diff
@@ -154,16 +154,38 @@ static int container_numindex(lua_State*L)
...
+    else
+    {
+      GtkWidget *searched_widget = g_list_nth_data(children, index);
+      lua_getuservalue(L, 1);
+      luaA_push(L, lua_widget, &searched_widget);
+    }
     g_list_free(children);
-    lua_getuservalue(L,1);
-    lua_pushlightuserdata(L,searched_widget);
-    lua_gettable(L,-2);
```
Comment (boucman) [inline src/lua/widget/container.c]: That last line has disapeared and that looks weird to me

the old code returned container->uservalue[searched_widget]  your new code returns searched_widget directely.

Could you check what's going on here ? IIRC because of the way GTK and lua interact, you need to go throught that gymnastic or you risk returning the gtk-widget instead of the lua widget...

File: src/lua/widget/combobox.c [hunk 5]
```diff
@@ -68,8 +68,6 @@ static int combobox_numindex(lua_State*L)
     return 0;
   }
   if(key <= 0 || key > length) {
-    return luaL_error(L,"Invalid index for combo box : %d\n",key);
-  } else if (key > length) {
```
Comment (theres) [inline src/lua/widget/combobox.c]: I think it's more consistent with regular lua behaviour - you can for example use ipair just like in regular table.

And this is the part of commit message for this:
> combobox fix: `for` with `ipairs` over combobox widget cause an error

However, I'll change this to be more accurate.

File: src/lua/widget/container.c (hunk 4 above)
Comment (theres) [inline src/lua/widget/container.c]: The old code doesn't work for me (see example file I prepared) - I just got `nil` instead lua_widget. But I'll check this once again.

File: src/lua/widget/container.c [hunk 6]
```diff
@@ -154,16 +154,38 @@ static int container_numindex(lua_State*L)
...
+        g_object_unref(current->data);
+      }
     }
     g_list_free(children);
     return 0;
   } else {
-    GtkWidget *searched_widget = g_list_nth_data(children,index);
+    if(index < 0 || index >= g_list_length(children))
+    {
+      lua_pushnil(L);
```
Comment (theres) [inline src/lua/widget/container.c]: and the same answer :) 

Even more, the old code for container doesn't raise an error, so behaviour of `container` and `combobox` was a bit inconsistent.

File: src/lua/widget/combobox.c (hunk 5 above)
Comment (boucman) [inline src/lua/widget/combobox.c]: wait... I don't see how this is related to ipairs, 

yes you are supposed to be able to iterate with ipairs, do you know why it didn't work , did ipair try to access out of bound element ?

maybe it would be enough to return nil for key == length+1 and an error for other access... 

I don't want to allow access to negative index (not sure exactly how lua reacts to that) and i'm a bit worried about accessing random beyond-end indices. If you set a random non-existing index, we would need to "fill up" the widget with the missing elements ?

File: src/lua/widget/container.c (hunk 6 above)
Comment (boucman) [inline src/lua/widget/container.c]: hmm, good point

**Comment by boucman** [comment]
Ok, your example file has convinced me, except maybe the pairs example

widgets should be consistant and i'm not sure how other widgets react to pairs() do they return their attributes ?

it would make sense that iterating on a widget would return all attributes (including childs) and ipairs would return child only..

File: src/lua/widget/container.c (hunk 4 above)
Comment (theres) [inline src/lua/widget/container.c]: Ok, so you are totally right, my code is working just by accident. I've found the problem: https://github.com/darktable-org/darktable/commit/0edd2f429b905e86ce5dbc6cbd3db1beea5d4c1f
You've changed `child_added`/`child_remove` handlers to setting: 
`container->uservalue[lua_widget] = lua_widget;`
so the `container->uservalue[searched_widget]` return nil.

**Comment by theres** [comment]
Iteration with `pairs` will return all key/value pairs (numeric + attributes). Ipairs start with 0 and try to retrive element at position greater by 1 until get `nil`. This is why it try to access out-of-bound element. It would be enough to return nil for `key == length+1` and an error for other access - I can do that this way if you want to, I just prefer current behaviour.

I've removed part of PR with containers children replacement, I'll add this in separate PR.

**Comment by LebedevRI** [comment]
@theres while not directly a feedback about these changes, but still
Please, write meaningful commit messages. First line should be under ~70 symbols,
and must fully describe the commit, and ideally should start with some prefix to
differentiate general area of the code that commit is about.
E.g. these commits should have a subject of something like 
```
Lua: container widget: fix access to children by index
```
If the commit messages are not meaningful, it is really painful to do cherry-picking to stable branch

**Comment by theres** [comment]
@LebedevRI I hope it's better

File: src/lua/widget/container.c [hunk 7]
```diff
@@ -88,7 +88,7 @@ static int child_added(lua_State *L)
   lua_widget widget;
   luaA_to(L, lua_widget,&widget,2);
   lua_getuservalue(L,1);
-  luaA_push(L,lua_widget,&widget);
+  lua_pushlightuserdata(L, widget->widget);
```
Comment (boucman) [inline src/lua/widget/container.c]: why change this ?

File: src/lua/widget/container.c [hunk 8]
```diff
@@ -99,7 +99,7 @@ static int child_removed(lua_State *L)
   lua_widget widget;
   luaA_to(L, lua_widget,&widget,2),
   lua_getuservalue(L,1);
-  luaA_push(L,lua_widget,&widget);
+  lua_pushlightuserdata(L, widget->widget);
```
Comment (boucman) [inline src/lua/widget/container.c]: same here

File: src/lua/widget/container.c (hunk 7 above)
Comment (theres) [inline src/lua/widget/container.c]: I'm pretty sure it's not functionnaly equivalent. Here we use raw `gtk_widget` instead of `lua_widget` as a key in uservalue. So as a result container uservalue contains mapping `gtk_widget -> lua_widget`. At the end we can easly look up for child lua_widget by gtk searched_widget, i.e. ` return container->user_value[searched_widget]`
 See the end of container_numindex function

```c
 static int container_numindex(lua_State*L)
 {
//...
   } else {
     GtkWidget *searched_widget = g_list_nth_data(children,index);
     g_list_free(children);
     lua_getuservalue(L,1);
     lua_pushlightuserdata(L,searched_widget);
     lua_gettable(L,-2);
     return 1;
   }
}
```  

So this is just a fix of the regression introduced by 0edd2f429b905e86ce5dbc6cbd3db1beea5d4c1f, I'm open to fix this other way if you have another proposition.

Comment (boucman) [inline src/lua/widget/container.c]: arn't these two pointers aliased via the type system , lua/widget.c line 102 aliases the gtk widget to the lua widget...

**Comment by TurboGit** [comment]
What is the status of this PR? Is that still needed/wanted?

Let me know, or I'll close this PR. Thanks.

**Comment by theres** [comment]
@TurboGit Imho yes.

@boucman could you take a look again?

**Comment by wpferguson** [comment]
@TurboGit I agree with @theres that this is needed.  I always end up coding around, or accepting the limitations, of the stack bug.

@theres, thanks for working on it.

**Comment by TurboGit** [comment]
Ok, in that case let this in. Thanks.
